docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,181 @@
+# Kubelet and systemd interaction
+
+**Author**: Derek Carr (@derekwaynecarr)
+
+**Status**: Proposed
+
+This document presents a specification for how the `kubelet` interfaces with `systemd`.
+
+## Motivation
+
+Many Linux distributions have either adopted, or plan to adopt `systemd` as their init system.
+
+This document describes a series of enhancements that should be made to the `kubelet` to better integrate with these distributions.
+
+## systemd units
+
+To help understand this proposal, we first provide a brief summary of `systemd` behavior.
+
+`systemd` manages a hierarchy of `slice`, `scope`, and `service` units.
+
+* `service` - application on the server that is launched by `systemd`; how it should start/stop; when it should be started; under what 
+circumstances it should be restarted; and any resource controls that should be applied to it.
+* `scope` - a process or group of processes which are not launched by `systemd` (i.e. fork), like a service, resource controls may be applied
+* `slice` - organizes a hierarchy in which `scope` and `service` units are placed.  a `slice` may contain `slice`, `scope`, or `service` units;
+processes are attached to `service` and `scope` units only, not to `slices`. The hierarchy is intended to be unified, meaning a process should
+only belong to a single leaf node.
+
+## cgroup hierarchy: orthogonal versus unified hierarchies and cgroupfs single writer
+
+Classical `cgroup` hierarchies were orthogonal per resource group controller, and a process may exist in different parts of the hierarchy.
+
+For example, a process `p1` could exist in each of the following at the same time:
+
+* `/sys/fs/cgroup/cpu/important/`
+* `/sys/fs/cgroup/memory/unimportant/`
+* `/sys/fs/cgroup/cpuacct/unimportant/`
+
+In addition, controllers for one resource group could depend on another in ways that were not always obvious.
+
+For example, the `cpu` controller depends on the `cpuacct` controller yet they were treated separately.
+
+Many found it confusing for a single process to belong to different nodes in the `cgroup` hierarchy across controllers.
+
+The Kernel direction for `cgroup` support is to move toward a unified `cgroup` hierarchy, where the per-controller hierarchies are eliminated in
+favor of hierarchies like the following:
+
+* `/sys/fs/cgroup/important/`
+* `/sys/fs/cgroup/unimportant/`
+
+In a unified hierarchy, a process should only belong to a single node in the `cgroup` tree.
+
+The Kernel direction for `cgroup` management is to promote a single-writer model rather than allowing multiple processes to independently
+write to parts of the file-system.
+
+## cgroup management with systemd and container runtimes
+
+By default, `systemd` defines three slices:
+
+* `system.slice` - contains system and application daemons
+* `user.slice` - user login related processes
+* `machine.slice` - virtual machines or containers that have init systems (not application space containers)
```

Comment:
@jonboulle - my concern was that I am not promoting kubernetes launching containers in `machine.slice` and without that clarification which the systemd docs try to state, then its confusing.  Do you have an opinion that it should be rooted in that slice?  Right now, `machine.slice` I view as the domain of things like libvirt.